Allow partial loading of nested fields

arbor-dwatson opened the pull request:
This change passes the `partial` keyword argument from a `schema.load`
call down into nested fields (see #438).  In the context of a REST
api, this allows a POST to require a subset of the fields to be
specified, while a PATCH can use `partial=True` to suppress any
required fields.

This change also supports using a tuple as the `partial` argument
using dot notation to configure partial loading of nested schemas.
For example: `partial=('x', 'y.z')` would allow `x` in the top level
schema to be missing or for `z` in the sub-schema for field `y` to be
missing.

I believe this is a breaking change since subclasses of `fields.Nested`
need to handle the `partial` parameter to `_deserialize()`.  This seems
to follow the existing pattern for using dot delimiters to specify nested
fields (`exclude` and `only` in the Schema init).  I'm very open to
suggestions on a better way to approach this.

arbor-dwatson committed:
Allow partial loading of nested fields

This change passes the `partial` keyword argument from a `schema.load`
call down into nested fields (see #438).  In the context of a REST
api, this allows a POST to require a subset of the fields to be
specified, while a PATCH can use `partial=True` to suppress any
required fields.

This change also supports using a tuple as the `partial` argument
using dot notation to configure partial loading of nested schemas.
For example: `partial=('x', 'y.z')` would allow `x` in the top level
schema to be missing or for `z` in the sub-schema for field `y` to be
missing.

arbor-dwatson committed:
Add myself to AUTHORS.rst.

arbor-dwatson committed:
Fix flake8 error.

I didn't follow the directions :(

lafrech commented on marshmallow/marshalling.py:
```diff
@@ -264,7 +265,18 @@ def deserialize(self, data, fields_dict, many=False, partial=False,
+            d_kwargs = {}
+            if isinstance(field_obj, fields.Nested):
+                # Allow partial loading of nested schemas.
+                prefix = field_name + "."
```
Minor optimization.

```py
                if partial_is_collection:
                    prefix = field_name + "."
                    len_prefix = len(prefix)
                    sub_partial = [f[len_prefix:]
                                   for f in partial if f.startswith(prefix)]
```

lafrech commented on tests/test_deserialization.py:
```diff
@@ -1431,6 +1431,77 @@ class Sch(Schema):
+        b_partial = ('z.x',)
+        result = SchemaB().load(b_dict, partial=b_partial)
+        assert result['z']['y'] == 42
+
```
Maybe add a failing test here to show the tuple is actually parsed, not just interpreted as `True`.

```py
        result = SchemaB().load(b_dict, partial=('z.x',))
        assert result['z']['y'] == 42

        with pytest.raises(ValidationError) as excinfo:
            SchemaB().load(b_dict, partial=('z.y',))
```

lafrech commented on marshmallow/fields.py:
```diff
@@ -306,6 +306,8 @@ def _deserialize(self, value, attr, data):
         :param value: The value to be deserialized.
         :param str attr: The attribute/key in `data` to be deserialized.
         :param dict data: The raw input data passed to the `Schema.load`.
+        :param bool|tuple partial: For nested schemas, the ``partial``
```
So, should ~either `partial` or~ `**kwargs` appear in the signature?

arbor-dwatson replied:
This feels like a project style decision so I didn't update it on my own.

My two cents: I think I can personally argue it both ways.  Adding it feels unnecessary from a client perspective.  It doesn't tell me anything about how I can directly use the function.  On the other hand, if I'm "using" the library by subclassing elements (which honestly feels like a typical use case), then I kind of want to know that the function will accept arbitrary keyword arguments.  The reason I didn't add it in the first place is that by the time I'm subclassing the function I'm already looking at the code so the documentation didn't seem as relevant.

lafrech commented:
Sorry @arbor-dwatson for the late reply. This looks good.

I guess we'd need to modify all the `_deserialize` signatures to add `**kwargs`.

Also, this implementation only passes `partial` (or subpartial tuple) to `Nested`, which makes sense and is more efficient. I'm wondering if it would be cleaner to pass it to any field, even though all the other fields will just ignore it. Singling out `Nested` prevents someone to implement his own nested field not deriving from `Nested`. But since `Nested` is already a special case everywhere in the code, it probably doesn't matter. I guess I'm overthinking it.

It just seems weird to modify all `_deserialize` signatures just for this special case, but it's not like it is breaking anything.

Also, adding `**kwargs` to all fields allows to add new parameters later on without it being a breaking change...

lafrech commented:
@sloria, @deckar01 feedback welcome. This is close to ready IMHO. I'm just not sure about the `_deserialize` signatures.

arbor-dwatson committed:
Add a negative test for a tuple partial argument.

arbor-dwatson committed:
Cache the prefix length (optimization).

arbor-dwatson commented:
@lafrech Thanks for the feedback!  I believe I've addressed your more technical concerns.  Looks like there are some conflicts with master that I'm going to ignore for now but I'm happy to resolve when the design decisions are finished.

sloria reviewed (approved):
@arbor-dwatson Sorry again for the late response on this. This looks great!

I don't feel strongly about adding `**kwargs` to all the field signatures, but I suppose it would be more future-proof. I don't think necessarily it needs to happen before this is merged, though.

For now, can you get this updated with `dev`, @arbor-dwatson ?

arbor-dwatson committed:
Merge branch 'dev' into partial-nested-schemas

arbor-dwatson committed:
Fix linting errors.

arbor-dwatson committed:
Fix bugs introduced by merge from dev.

 - Somehow I lost the **kwargs in Nested _deserialize.
 - Added **kwargs to Pluck(Nested) _deserialize.

arbor-dwatson commented:
@sloria No problem with the delay!  Thank you for all your hard work!

Updated with dev.  The only real issue was the new Pluck class, which I think I updated correctly.

Let me know if there's anything else I can do.

sloria commented:
Thanks for updating it. Looks good to merge.
